dev/TEAL.md

```diff
@@ -0,0 +1,219 @@
+| 15 | XferAsset | []byte |
+| 16 | AssetAmount | uint64 |
+| 17 | AssetSender | []byte |
+| 18 | AssetReceiver | []byte |
+| 19 | AssetCloseTo | []byte |
+| 20 | GroupIndex | uint64 |
+
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Round | uint64 |
+| 1 | MinTxnFee | uint64 |
+| 2 | MinBalance | uint64 |
+| 3 | MaxTxnLife | uint64 |
+| 4 | TimeStamp | uint64 |
+| 5 | ZeroAddress | []byte |
+| 6 | GroupSize | uint64 |
+
+
+
+### Flow Control
+
+| Op | Description |
+| --- | --- |
+| `err` | Error. Panic immediately. This is primarily a fencepost against accidental zero bytes getting compiled into programs. |
+| `bnz` | branch if value is not zero |
+| `pop` | discard value from stack |
+| `dup` | duplicate last value on stack |
+
+# Assembler Syntax
+
+The assembler parses line by line. Ops that just use the stack appear on a line by themselves. Ops that take arguments are the op and then whitespace and then any argument or arguments.
+
+"`//`" prefixes a line comment.
+
+## Constants and Pseudo-Ops
+
+A few pseudo-ops simplify writing code. `int` and `byte` and `addr` followed by a constant record the constant to a `intcblock` or `bytecblock` at the beginning of code and insert an `intc` or `bytec` reference where the instruction appears to load that value. `addr` parses an Algorand account address base32 and converts it to a regular bytes constant.
+
+`byte` constants are:
+```
+byte base64 AAAA...
+byte b64 AAAA...
+byte base64(AAAA...)
+byte b64(AAAA...)
+byte base32 AAAA...
+byte b32 AAAA...
+byte base32(AAAA...)
+byte b32(AAAA...)
+byte 0x0123456789abcdef...
+```
+
+`int` constants may be `0x` prefixed for hex, `0` prefixed for octal, or decimal numbers.
+
+`intcblock` may be explictily assembled. It will conflict with the assembler gathering `int` pseudo-ops into a `intcblock` program prefix, but may be used in code only has explicit `intc` references. `intcblock` should be followed by space separated int constants all on one line.
+
+`bytecblock` may be explicitly assembled. It will conflict with the assembler if there are any `byte` pseudo-ops but may be used if only explicit `bytec` references are used. `bytecblock` should be followed with byte contants all on one line, either 'encoding value' pairs (`b64 AAA...`) or 0x prefix or function-style values (`base64(...)`).
+
+## Labels and Branches
+
+A label is defined by any string not some other op or keyword and ending in ':'. A label can be an argument (without the trailing ':') to a branch instruction.
+
+Example:
+```
+int 1
+bnz safe
+err
+safe:
+pop
+```
+
+# Encoding and Versioning
+
+A program starts with a varuint declaring the version of the compiled code. Any addition, removal, or change of opcode behavior increments the version. For the most part opcode behavior should not change, addition will be infrequent (not likely more often than every three months and less often as the language matures), and removal should be very rare.
```

Comment:
If an old opcode is removed for whatever reason, do old programs break?

Reply:
We'll need provisions for running old version programs as-if they were still in that environment. Op behavior changes need to be versioned. Transaction fields need to be versioned.